Explore: Adds support for new loki 'start' and 'end' params for labels endpoint

kaydelaney opened the pull request:
**What this PR does / why we need it**:
Fixes an issue on Loki explore where labels would disappear after ingestion was flushed since labels were only queried from ingestors.

**Which issue(s) this PR fixes**:
Closes #16788

**Special notes for your reviewer**:
See:
grafana/loki#453
grafana/loki#521

hugohaggmark reviewed (commented):
Wow, great work! I talked to @marefr during lunch and he has some comments about the `range` that I totally forgot about.

marefr commented:
We have a sort of design flaw in Explore current when you have selected a relative time range, e.g. last 6 hours, you actually have to parse the raw time range before getting the correct absolute time range in epoch ms.

Currently when running queries, we make sure to calculate the time range in epoch ms before executing the queries:
https://github.com/grafana/grafana/blob/40f95a95c011199e7fa7f69d22768480c56680a6/public/app/features/explore/state/actions.ts#L399-L411

So I guess you have to do similar thing as in row 406-407 before being able to use the range. Does this all makes sense?

kaydelaney commented:
@marefr Yep, makes sense! Will fix now.

marefr commented:
@kaydelaney yes please try that out. It's related to #17629

torkelo commented on public/app/plugins/datasource/loki/language_provider.ts:
```diff
@@ -120,7 +123,7 @@ export default class LokiLanguageProvider extends LanguageProvider {
   }
 
-  getLabelCompletionItems({ text, wrapperClasses, labelKey, value }: TypeaheadInput): TypeaheadOutput {
+  getLabelCompletionItems({ text, wrapperClasses, labelKey, value }: TypeaheadInput, { range }: any): TypeaheadOutput {
```
can't see an update call to this function, where is this called with range passed in?

kaydelaney replied:
See Line 91:
```javascript
 return this.getLabelCompletionItems.apply(this, arguments);
```
and then in `LokiQueryFieldForm.tsx`:
```javascript
    const result = datasource.languageProvider.provideCompletionItems(
      { text, value, prefix, wrapperClasses, labelKey },
      { history, range }
    );
```

torkelo commented on public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx:
```diff
@@ -70,6 +70,7 @@ export interface LokiQueryFieldFormProps extends ExploreQueryFieldProps<DataSour
   syntax: any;
   logLabelOptions: any[];
   syntaxLoaded: any;
+  range: TimeRange;
```
If we are to add this to the query editor interface it should be added to the base QueryEditorProps interface (and property should be named timeRange so it's named the same as in PanelProps interface)

marefr replied:
@torkelo we're using a property named `absoluteRange` of type `AbsoluteRange` now. Still feel this should go into `QueryEditorProps` interface?

torkelo replied:
maybe it's not needed, the range is also available on the last query request/response

marefr replied:
Okay so let's move on with this and we may look further into this later.

marefr commented:
How does this implementation feel to you @kaydelaney. Would be interesting to get some comments from the developer point of view. Seems like we're pushing and pushing the range and time zone lower and lower than thru the component-chain and that each datasource/language provider needs to implement "parsing" of time range. 

Maybe it's possible to add some behavior to base classes/components of `QueryField` or `LanguageProvider`? Or alternatively pass down a function to retrieve the parsed time range upon request, guess this makes it similar to `timeSrv`. Or have some sort of global time range, but it's tricky since you may have a left and a right explore with different time ranges.

marefr commented:
Should be able to rebase and continue with this as soon as #17677 is merged.

marefr reviewed (changes requested):
FYI the `timeSrv` is now updated as part of #17677 so as a first step we may not need to pass the range to metadataRequest and use `timeSrv` instead.

If you still think that passing the range to metadataRequest is the way to go: A new property `absoluteRange` was introduced in #17677 that I would like you to use instead of passing both a `TimeRange` and a `TimeZone` as props downwards.

kaydelaney committed:
Explore: Adds support for new loki 'start' and 'end' params for labels endpoint
Also initializes absoluteRange when explore is initialized
Closes #16788

marefr reviewed (changes requested):
Some initial feedback. Still haven't tested this, but planning to do so.

kaydelaney committed:
Explore: Dispatches updateTimeRangeAction instead of passing absoluteRange when initializing
Also removes dependency on sinon from loki language provider test

kaydelaney committed:
Loki: Refactors transformation of absolute time range to URL params into small utility function

marefr commented on public/app/plugins/datasource/loki/language_provider.test.ts:
```diff
@@ -128,8 +128,8 @@ describe('Request URL', () => {
     const expectedParams = {
-      start: rangeMock.from.valueOf() * NS_IN_MS,
-      end: rangeMock.to.valueOf() * NS_IN_MS,
+      start: rangeMock.from * NS_IN_MS,
```
Could use the `rangeToParams` here as well and maybe in other places?

marefr commented on public/app/plugins/datasource/loki/language_provider.ts:
```diff
@@ -67,13 +71,13 @@ export default class LokiLanguageProvider extends LanguageProvider {
 
-  request = (url: string) => {
-    return this.datasource.metadataRequest(url);
+  request = (url: string, params?: any) => {
```
The request function of abstract class `LanguageProvider` in types/Explore.ts should be updated with `params` argument and probably string as `url` type.

marefr reviewed (changes requested):
Great fixes of my last comments :+1: Added some minor comments.

Spotted one odd thing. When going to Explore and loading Loki I see a label request ending in `?` and no start/end, see request below. Not sure if this is intended? As soon as I click on Log Labels after this requests for log labels (with start/end) and label values (with start/end) are included.

![image](https://user-images.githubusercontent.com/1668778/60630621-33a03400-9dfb-11e9-80a3-e03a645dcfc2.png)

Reloading the whole page gives me the following requests, where one of the label requests also ends in `?` and has no start/end params. Again, not sure this is intended or not?

![image](https://user-images.githubusercontent.com/1668778/60630838-53842780-9dfc-11e9-9463-3be81bfee32b.png)

kaydelaney committed:
Makes use of rangeToParams() util function in loki language provider test
Also updates LanguageProvider.request() interface so that url should be type string, and adds optional params argument

kaydelaney commented:
@marefr regarding the label request with no start/end params, that's just the request being made in `testDatasource()` to verify that the data source can be connected to and that labels can be retrieved. The data from the response doesn't get used so there shouldn't be any need to pass the parameters, I think.

marefr reviewed (changes requested):
Looks good :+1: Just fix the conflicts and I'll approve this

marefr reviewed (approved):
LGTM

davkal commented:
Just tested this too. Works well 🎉
